qutip/core/operators.py

```diff
@@ -102,166 +109,181 @@ def jmat(j, which=None):
         raise ValueError('j must be a non-negative integer or half-integer')
 
     if not which:
-        return jmat(j, 'x'), jmat(j, 'y'), jmat(j, 'z')
+        return (
+            jmat(j, 'x', dtype=dtype),
+            jmat(j, 'y', dtype=dtype),
+            jmat(j, 'z', dtype=dtype)
+        )
 
     dims = [[int(2*j + 1)]]*2
     if which == '+':
-        return Qobj(_jplus(j), dims=dims, type='oper',
+        return Qobj(_jplus(j, dtype=dtype), dims=dims, type='oper',
                     isherm=False, isunitary=False, copy=False)
     if which == '-':
-        return Qobj(_jplus(j).adjoint(), dims=dims, type='oper',
+        return Qobj(_jplus(j, dtype=dtype).adjoint(), dims=dims, type='oper',
                     isherm=False, isunitary=False, copy=False)
     if which == 'x':
-        A = 0.5 * _jplus(j)
+        A = 0.5 * _jplus(j, dtype=dtype)
         return Qobj(A + A.adjoint(), dims=dims, type='oper',
                     isherm=True, isunitary=False, copy=False)
     if which == 'y':
-        A = -0.5j * _jplus(j)
+        A = -0.5j * _jplus(j, dtype=dtype)
         return Qobj(A + A.adjoint(), dims=dims, type='oper',
                     isherm=True, isunitary=False, copy=False)
     if which == 'z':
-        return Qobj(_jz(j), dims=dims, type='oper',
+        return Qobj(_jz(j, dtype=dtype), dims=dims, type='oper',
                     isherm=True, isunitary=False, copy=False)
     raise ValueError('invalid spin operator: ' + which)
 
 
-def _jplus(j):
+def _jplus(j, *, dtype=None):
     """
     Internal functions for generating the data representing the J-plus
     operator.
     """
+    dtype = to.str2type(dtype or _data.CSR)
```

Comment:
I'd have `str2type` be private to `data.to`, and just have `__getitem__` call it.  The fewer public functions our objects have, the easier they are to use.

It also means that we can guaranteed that the interface is uniform _everywhere_, and isn't dependent on us writing additional boilerplate within each individual function.

Reply:
This comment recurs on the other ~20 lines it appears within this file.

Reply:
Maybe we'd keep the approximate functionality, but call it `to.parse` and have it work only on scalar inputs?  Then it _can_ be used externally, but in general I would expect `Dispatcher.__getitem__` to call it internally.

Reply:
I preferred not to add any extra step to the `Dispatcher`, keeping the string `dtype` as a user format only, having any user provided `dtype` converted as pure type immediately and keeping all internal with only that format.
(For me anything that works with Data-layer directly is internal.)
This is why, I made `str2type` available to the python layer.
I will change it to `to.parse` and leave the `Dispatcher` untouched for now.

Reply:
But this is an extra step to Dispatcher - it's just that the complexity of the function is now spread around in many many places, which leaves us open to forgetting to do it at some point.  Putting this test in `Dispatcher` doesn't add any additional runtime, and it makes a single source of truth for "what is a data-layer type".  That's much easier to maintain and extend.

Reply:
`expm` takes a `dtype` input, under this logic should it be parsed by the `Dispatcher`? 
If yes, I am not sure I will get all place where I should parse in `Dispatcher`...
I will do the change in `Dispatcher.__getitem__`, `Dispatcher.__call__` , but I am pretty sure I am missing some places...